### DISCUSSION TITLE
Is it possible to mix UploadFile with List of UploadFile in an endpoint?

### DISCUSSION THREAD

**Discussion by guzman109** [fastapi/fastapi#12051] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@router.post("/upload",status_code=status.HTTP_201_CREATED)
def upload(
   file_a: Annotated[UploadFile, File(description="File A")],
   file_b: Annotated[UploadFile | None, File(description="Optional File B")] = None,
   files_extra: Annotated[list[UploadFile] | None, File(description="Optional Extra Files")] = None
):
 #
```

### Description

I'm trying to write an endpoint that where the first file is mandatory.
The second is optional but if uploaded,  I need to make sure its the correct type of file. 
Other files can be uploaded as well to store and I  just store then as is. (no verification).

I'm using the following to test the endpoint

Using httpx,
```python
files = [
   ( "file_a", ("file_a.txt", open("file_a.txt", "rb"), "text/plain") ),
   ( "file_b", ("file_b.txt", open("file_b.txt", "rb"), "text/plain") )
   [
         ( "files_extra", ("file_c.txt", open("file_c.txt", "rb"), "text/plain") ),
         ( "files_extra", ("file_d.txt", open("file_d.txt", "rb"), "text/plain") ),
         ( "files_extra", ("file_e.txt", open("file_e.txt", "rb"), "text/plain") )
   ]
]
response = httpx.post(url, files=files)
response.raise_for_status()
```

But I get a 422 Error and this 
```python
{'detail': [{'type': 'list_type',
   'loc': ['body', 'other_files'],
   'msg': 'Input should be a valid list',
   'input': {'filename': 'file_c.txt',
    'file': {'_file': {},
     '_max_size': 1048576,
     '_rolled': False,
     '_TemporaryFileArgs': {'mode': 'w+b',
      'buffering': -1,
      'suffix': None,
      'prefix': None,
      'encoding': None,
      'newline': None,
      'dir': None,
      'errors': None}},
    'size': 1408,
    'headers': {'content-disposition': 'form-data; name="files_extra"; filename="file_c.txt"',
     'content-type': 'text/plain'}}}]}
```

### Operating System

Linux

### FastAPI Version

0.112.0

### Pydantic Version

2.8.2

### Python Version

3.12.3

**Comment by iamdhakrey** [comment, marked as answer]
@guzman109 you need to replace `files_extra: Annotated[list[UploadFile] | None, File(description="Optional Extra Files")] = None` to `files_extra: Annotated[List[UploadFile], File(description="Optional Extra Files")] = []` because the combination of List[UploadFile] | None with File(...) creates a conflict. FastAPI expects a required file list due to File(...), but if it receives None (as might be the case when files are not provided), it leads to processing issues.

```diff
files = [
   ( "file_a", ("file_a.txt", open("file_a.txt", "rb"), "text/plain") ),
   ( "file_b", ("file_b.txt", open("file_b.txt", "rb"), "text/plain") )
-  [
+ # Each 'files_extra' is treated as a separate tuple, but they all share the same key
         ( "files_extra", ("file_c.txt", open("file_c.txt", "rb"), "text/plain") ),
         ( "files_extra", ("file_d.txt", open("file_d.txt", "rb"), "text/plain") ),
         ( "files_extra", ("file_e.txt", open("file_e.txt", "rb"), "text/plain") )
-  ]
]
```

**Reply by guzman109** [reply]
Thank you! All working now.
